# Data wrangling

## 1-Imports

In [19]:
#Import pandas
import pandas as pd
from library.sb_utils import save_file

## 2-Load the data


In [20]:
# Load The meal delivery company Data: Food Orders, Fulfilment center info, and meals info
orders = pd.read_csv('../Data/Food_orders_data.csv')
centers = pd.read_csv('../Data/fulfilment_center_info.csv')
meals= pd.read_csv('../Data/meal_info.csv')

## 3-Explore the data

In [21]:
#use the info method to see a summary of the data
orders.info()
centers.info()
meals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 423727 entries, 0 to 423726
Data columns (total 9 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id                     423727 non-null  int64  
 1   week                   423727 non-null  int64  
 2   center_id              423727 non-null  int64  
 3   meal_id                423727 non-null  int64  
 4   checkout_price         423727 non-null  float64
 5   base_price             423727 non-null  float64
 6   emailer_for_promotion  423727 non-null  int64  
 7   homepage_featured      423727 non-null  int64  
 8   num_orders             423727 non-null  int64  
dtypes: float64(2), int64(7)
memory usage: 29.1 MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77 entries, 0 to 76
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   center_id    77 non-null     int64  
 1   city_code    77 non-nul

In [22]:
#use the head method to see the first several rows of the data
orders.head()

id  week  center_id  meal_id  checkout_price  base_price  \
0  1379560     1         55     1885          136.83      152.29   
1  1466964     1         55     1993          136.83      135.83   
2  1346989     1         55     2539          134.86      135.86   
3  1338232     1         55     2139          339.50      437.53   
4  1448490     1         55     2631          243.50      242.50   

   emailer_for_promotion  homepage_featured  num_orders  
0                      0                  0         177  
1                      0                  0         270  
2                      0                  0         189  
3                      0                  0          54  
4                      0                  0          40

In [23]:
centers.head()

center_id  city_code  region_code center_type  op_area
0         11        679           56      TYPE_A      3.7
1         13        590           56      TYPE_B      6.7
2        124        590           56      TYPE_C      4.0
3         66        648           34      TYPE_A      4.1
4         94        632           34      TYPE_C      3.6

In [24]:
meals.head()

meal_id   category cuisine
0     1885  Beverages    Thai
1     1993  Beverages    Thai
2     2539  Beverages    Thai
3     1248  Beverages  Indian
4     2631  Beverages  Indian

In [25]:
#join all three tables
food_data=pd.merge(pd.merge(orders,centers, on='center_id'),meals, on='meal_id')
food_data.head()

id  week  center_id  meal_id  checkout_price  base_price  \
0  1379560     1         55     1885          136.83      152.29   
1  1018704     2         55     1885          135.83      152.29   
2  1196273     3         55     1885          132.92      133.92   
3  1116527     4         55     1885          135.86      134.86   
4  1343872     5         55     1885          146.50      147.50   

   emailer_for_promotion  homepage_featured  num_orders  city_code  \
0                      0                  0         177        647   
1                      0                  0         323        647   
2                      0                  0          96        647   
3                      0                  0         163        647   
4                      0                  0         215        647   

   region_code center_type  op_area   category cuisine  
0           56      TYPE_C      2.0  Beverages    Thai  
1           56      TYPE_C      2.0  Beverages    Thai  
2           56      TYPE_C      2.0  Beverages    Thai  
3           56      TYPE_C      2.0  Beverages    Thai  
4           56      TYPE_C      2.0  Beverages    Thai

In [26]:
food_data

id  week  center_id  meal_id  checkout_price  base_price  \
0       1379560     1         55     1885          136.83      152.29   
1       1018704     2         55     1885          135.83      152.29   
2       1196273     3         55     1885          132.92      133.92   
3       1116527     4         55     1885          135.86      134.86   
4       1343872     5         55     1885          146.50      147.50   
...         ...   ...        ...      ...             ...         ...   
423722  1282954   131         61     2104          445.29      630.53   
423723  1094245   132         61     2104          435.53      630.53   
423724  1380866   133         61     2104          601.43      630.53   
423725  1111394   134         61     2104          600.43      631.53   
423726  1439333   135         61     2104          602.43      631.53   

        emailer_for_promotion  homepage_featured  num_orders  city_code  \
0                           0                  0         177        647   
1                           0                  0         323        647   
2                           0                  0          96        647   
3                           0                  0         163        647   
4                           0                  0         215        647   
...                       ...                ...         ...        ...   
423722                      0                  1         270        473   
423723                      0                  1         339        473   
423724                      0                  1         107        473   
423725                      0                  1          41        473   
423726                      0                  1          28        473   

        region_code center_type  op_area   category      cuisine  
0                56      TYPE_C      2.0  Beverages         Thai  
1                56      TYPE_C      2.0  Beverages         Thai  
2                56      TYPE_C      2.0  Beverages         Thai  
3                56      TYPE_C      2.0  Beverages         Thai  
4                56      TYPE_C      2.0  Beverages         Thai  
...             ...         ...      ...        ...          ...  
423722           77      TYPE_A      4.5       Fish  Continental  
423723           77      TYPE_A      4.5       Fish  Continental  
423724           77      TYPE_A      4.5       Fish  Continental  
423725           77      TYPE_A      4.5       Fish  Continental  
423726           77      TYPE_A      4.5       Fish  Continental  

[423727 rows x 15 columns]

In [27]:
food_data=food_data.sort_values(by='week',ascending=True)
food_data

id  week  center_id  meal_id  checkout_price  base_price  \
0       1379560     1         55     1885          136.83      152.29   
128062  1044078     1         39     2290          301.73      299.73   
127927  1272063     1         65     2290          305.58      306.58   
127792  1364972     1         43     2290          279.39      279.39   
127657  1363806     1         80     2290          311.43      311.43   
...         ...   ...        ...      ...             ...         ...   
235006  1054964   135        152     2581          533.53      532.53   
235141  1360246   135         97     2581          533.53      549.08   
235274  1316963   135         74     2581          532.53      532.53   
232475  1302921   135         86     1558          548.08      640.23   
423726  1439333   135         61     2104          602.43      631.53   

        emailer_for_promotion  homepage_featured  num_orders  city_code  \
0                           0                  0         177        647   
128062                      0                  0         419        526   
127927                      0                  0         878        602   
127792                      0                  0         944        590   
127657                      0                  0         446        604   
...                       ...                ...         ...        ...   
235006                      1                  0         176        576   
235141                      1                  1         311        628   
235274                      1                  0          81        702   
232475                      1                  1         717        699   
423726                      0                  1          28        473   

        region_code center_type  op_area   category      cuisine  
0                56      TYPE_C      2.0  Beverages         Thai  
128062           34      TYPE_C      3.8  Rice Bowl       Indian  
127927           34      TYPE_A      4.8  Rice Bowl       Indian  
127792           56      TYPE_A      5.1  Rice Bowl       Indian  
127657           56      TYPE_C      5.1  Rice Bowl       Indian  
...             ...         ...      ...        ...          ...  
235006           34      TYPE_B      4.0      Pizza  Continental  
235141           77      TYPE_A      4.6      Pizza  Continental  
235274           35      TYPE_A      2.8      Pizza  Continental  
232475           85      TYPE_C      4.0      Pizza  Continental  
423726           77      TYPE_A      4.5       Fish  Continental  

[423727 rows x 15 columns]

In [28]:
#change 'category' column name 
food_data.rename(columns={'category':'meal_category'},inplace=True)

In [29]:
# Save the data 

datapath = '../data'
save_file(food_data, 'food_data_merged.csv', datapath)

Writing file.  "../data/food_data_merged.csv"


In [30]:
food_data.head()

id  week  center_id  meal_id  checkout_price  base_price  \
0       1379560     1         55     1885          136.83      152.29   
128062  1044078     1         39     2290          301.73      299.73   
127927  1272063     1         65     2290          305.58      306.58   
127792  1364972     1         43     2290          279.39      279.39   
127657  1363806     1         80     2290          311.43      311.43   

        emailer_for_promotion  homepage_featured  num_orders  city_code  \
0                           0                  0         177        647   
128062                      0                  0         419        526   
127927                      0                  0         878        602   
127792                      0                  0         944        590   
127657                      0                  0         446        604   

        region_code center_type  op_area meal_category cuisine  
0                56      TYPE_C      2.0     Beverages    Thai  
128062           34      TYPE_C      3.8     Rice Bowl  Indian  
127927           34      TYPE_A      4.8     Rice Bowl  Indian  
127792           56      TYPE_A      5.1     Rice Bowl  Indian  
127657           56      TYPE_C      5.1     Rice Bowl  Indian

In [31]:
#use the info method to see a summary of the data
food_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 423727 entries, 0 to 423726
Data columns (total 15 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id                     423727 non-null  int64  
 1   week                   423727 non-null  int64  
 2   center_id              423727 non-null  int64  
 3   meal_id                423727 non-null  int64  
 4   checkout_price         423727 non-null  float64
 5   base_price             423727 non-null  float64
 6   emailer_for_promotion  423727 non-null  int64  
 7   homepage_featured      423727 non-null  int64  
 8   num_orders             423727 non-null  int64  
 9   city_code              423727 non-null  int64  
 10  region_code            423727 non-null  int64  
 11  center_type            423727 non-null  object 
 12  op_area                423727 non-null  float64
 13  meal_category          423727 non-null  object 
 14  cuisine                423727 non-null  o

The selected data types for the columns seem appropriate.

In [32]:
#Perform summary statistics on numeric features
food_data.describe()

id           week      center_id        meal_id  \
count  4.237270e+05  423727.000000  423727.000000  423727.000000   
mean   1.249964e+06      69.677332      82.109469    2023.956295   
std    1.443431e+05      38.686607      45.977420     547.370379   
min    1.000000e+06       1.000000      10.000000    1062.000000   
25%    1.124878e+06      37.000000      43.000000    1558.000000   
50%    1.249955e+06      70.000000      76.000000    1971.000000   
75%    1.374949e+06     103.000000     110.000000    2539.000000   
max    1.499999e+06     135.000000     186.000000    2956.000000   

       checkout_price     base_price  emailer_for_promotion  \
count   423727.000000  423727.000000          423727.000000   
mean       331.526166     353.732761               0.083353   
std        153.085823     160.919513               0.276416   
min          2.970000      55.350000               0.000000   
25%        228.010000     243.500000               0.000000   
50%        292.030000     309.490000               0.000000   
75%        444.290000     458.870000               0.000000   
max        866.270000     866.270000               1.000000   

       homepage_featured     num_orders      city_code    region_code  \
count      423727.000000  423727.000000  423727.000000  423727.000000   
mean            0.111315     264.550536     601.562478      56.609468   
std             0.314522     403.573030      66.206777      17.641225   
min             0.000000      13.000000     456.000000      23.000000   
25%             0.000000      54.000000     553.000000      34.000000   
50%             0.000000     136.000000     596.000000      56.000000   
75%             0.000000     324.000000     651.000000      77.000000   
max             1.000000   24299.000000     713.000000      93.000000   

             op_area  
count  423727.000000  
mean        4.083245  
std         1.091407  
min         0.900000  
25%         3.600000  
50%         4.000000  
75%         4.500000  
max         7.000000

We want to predict the number of orders for each meal at each center. So, the target variable is number of orders (num_orders).

### Check for missing data

In [33]:
# check if there is any missing values
food_data.isna().sum()

id                       0
week                     0
center_id                0
meal_id                  0
checkout_price           0
base_price               0
emailer_for_promotion    0
homepage_featured        0
num_orders               0
city_code                0
region_code              0
center_type              0
op_area                  0
meal_category            0
cuisine                  0
dtype: int64

There is no missing data in the dataset.

#### Features unique values

In [34]:
#Estimate number of unique values for some features
meals=food_data.meal_id.unique().size
centers=food_data.center_id.unique().size
city=food_data.city_code.unique().size
region=food_data.region_code.unique().size
centerType=food_data.center_type.unique().size
meal_cat=food_data.meal_category.unique().size
cuisine=food_data.cuisine.unique().size

print(f'center={centers}, meal={meals}, city= {city}, region= {region}, center_Type= {centerType}, meal_cat= {meal_cat}, cuisine= {cuisine}')

center=77, meal=51, city= 51, region= 8, center_Type= 3, meal_cat= 14, cuisine= 4


In [35]:
food_data.emailer_for_promotion.value_counts()

emailer_for_promotion
0    388408
1     35319
Name: count, dtype: int64

In [36]:
food_data.homepage_featured.value_counts()

homepage_featured
0    376560
1     47167
Name: count, dtype: int64

#### Checking the checkout price and base price to explore their relationship


In [37]:
checkout=len(food_data[food_data['checkout_price']>food_data['base_price']])
checkout

107102

In [38]:
base=len(food_data[food_data['checkout_price']<food_data['base_price']])
base

214227

In [39]:
check_base=len(food_data[food_data['checkout_price']==food_data['base_price']])
check_base

102398

The above calculation suggests that the checkout price varies in relation to the base price, with a substantial number of cases where the checkout price is either greater than or less than the base price. The almost equal count for cases where the checkout price equals the base price (102,398) further supports the idea that there isn't a specific consistent relationship between the checkout and base prices.

#### Checking the email promotion and page featured and thier relationship

In [40]:
    (food_data[food_data.emailer_for_promotion!= food_data.homepage_featured]
 .groupby('emailer_for_promotion')['homepage_featured']
 .value_counts())

emailer_for_promotion  homepage_featured
0                      1                    28873
1                      0                    17025
Name: count, dtype: int64

In [41]:
    (food_data[food_data.emailer_for_promotion== food_data.homepage_featured]
 .groupby('emailer_for_promotion')['homepage_featured']
 .value_counts())

emailer_for_promotion  homepage_featured
0                      0                    359535
1                      1                     18294
Name: count, dtype: int64

There seems to be a correlation between email promotions and homepage features. When they are not featured together, there are fewer instances. However, when they are featured together or not featured together, the counts are higher, indicating a potential relationship or alignment between these two features.